Analytics: unify calypso_user_registration_complete events and move identifyUser() inside recordRegistration()

MicBosi opened the pull request:
#### Changes proposed in this Pull Request

1) Unifies `calypso_user_registration_complete` events into one.
2) Moves `identifyUser()` inside the one `recordRegistration()` thus making sure `calypso_user_registration_complete` is already aliased and `identifyUser()` happens no matter what new signup method will be added in the future.
3) A few more cosmetic fixes reported by either CI or the JS linter.

#### Testing instructions

* Enabled trackers and logs:
```
document.cookie = 'flags=gdpr-banner,google-analytics,ad-tracking; path=/; domain=.' + document.location.hostname.split( '.' ).slice( -2 ).join( '.' );
document.cookie = 'sensitive_pixel_option=yes; path=/; domain=.' + document.location.hostname.split( '.' ).slice( -2 ).join( '.' );
localStorage.setItem( 'debug', 'calypso:analytics*' );
```

* Test normal registration, passwordless registration, social registration
* Make sure `flow` and `type` props are present and properly set in the `calypso_user_registration_complete` Tracks event.
* Make sure there are no regressions with https://github.com/Automattic/wp-calypso/pull/37263

p2-pbmxuV-U

MicBosi committed:
Analytics: unify 'calypso_user_registration_complete' events

MicBosi committed:
Move identifyUser() inside and at the top of recordRegistration()

jaswrks commented on client/lib/analytics/index.js:
```diff
@@ -345,33 +345,16 @@ const analytics = {
 
-	recordRegistration: function( { flow } ) {
+	recordRegistration: function( { username, userId, flow, type } ) {
+		analytics.identifyUser( username, userId );
```
I see a couple of merge conflicts in this branch, but looks very good overall. I think once you resolve conflicts you'll see that we need this call to `identifyUser()` to include the email address too, and it expects an object now.

MicBosi replied:
Thanks @jaswrks this is done now.

andrewserong commented:
This is looking great to me, thanks for putting it together. I don't have anything else to add beyond what @jaswrks mentioned, and the default and passwordless signups are working nicely, passing in the correct `type` property.

adamziel commented:
@MicBosi @jaswrks what can we do to help merge this one?

MicBosi committed:
Pass user object to identifyUser()

MicBosi committed:
Fixed CI errors complaining about unknown global vars `Image`, `btoa`, `location`

MicBosi committed:
Use standard `window.location` instead of legacy `document.location`, fixed lint errors

MicBosi commented:
Thanks @andrewserong and @adamziel I've fixed the conflicts and rebased the code against the new `identifyUser()` signature. Then a few CI issues came up due to undefined globals and other linter related errors which I had to fix in order to commit. Haven't fully tested this yet but if you could have another look that would be great.

I've also added a few debug lines in the top level functions of `analytics`. Setting `localStorage.setItem( 'debug', 'calypso:analytics*' );` should result in showing them and it makes it easier to inspect the user data, flow etc. upon registration.

cc @jaswrks

andrewserong commented on client/lib/signup/step-actions.js:
```diff
@@ -415,7 +415,24 @@ export function createAccount(
+				const registrationUserData = {
+					ID: userId,
+					username: username,
+					email: userData.email
```
This throws an error in social signup for me (tested in a local dev environment using the `Continue with Google` button). It looks like for [Google signups](https://github.com/automattic/wp-calypso/blob/f2c9d99b07e14ea938829e9b2e5b31b6fb388a90/client/blocks/signup-form/social.jsx#L55) no additional userData is passed in so this will be `null`. And for [Apple signups](https://github.com/automattic/wp-calypso/blob/f2c9d99b07e14ea938829e9b2e5b31b6fb388a90/client/blocks/signup-form/social.jsx#L40), the property is `user_email` instead of `email` — I haven't had a chance to test signing in with Apple, but I suspect this will throw the same error. Perhaps the Apple one could be renamed? For Google, it looks like there should be an email address in the response object at `response.email`, I think?

andrewserong replied:
Error message was `step-actions.js:421 Uncaught TypeError: Cannot read property 'email' of null`

andrewserong reviewed (changes requested):
This is testing well for standard signup and passwordless signup, thanks for rebasing and continuing on with this one! Just ran into an error with the social signup, and I've added a note.

I haven't found any other regressions related to the PR you mentioned, but @jaswrks might have a better idea about that side of things.

MicBosi commented:
Fix moved to this other PR: https://github.com/Automattic/wp-calypso/pull/40446
